### ISSUE TITLE
Support DNS over TCP

### ISSUE THREAD

**Issue by Smittyvb** [tailscale/tailscale#1123]
**Describe the bug**
Magic DNS doesn't support TCP, preventing queries with long records from working.

**To Reproduce**
```bash
# Works, outputs long record: (UDP request, TCP response)
dig @8.8.8.8 -t TXT longrecord.smitop.com

# No record and "Message parser reports malformed message packet": (UDP request, TCP response)
dig @100.100.100.100 -t TXT longrecord.smitop.com

# Doesn't even go through the Magic DNS system, "open-conn-track: timeout opening (100.66.99.116:48353 => 100.100.100.100:53); no associated peer node" (TCP request, TCP response)
dig @100.100.100.100 +tcp smitop.com
```

**Expected behavior**
Magic DNS dosen't support DNS over TCP, currently it only supports UDP. As such, long records that don't fit in a UDP packet fail (even for IPs on the public Internet). It at least shouldn't return malformed packets. This also causes issues with DNSSEC, where answer sizes are larger due to being signed (not a security issue though since I believe failure here would just cause the answer to be rejected))

**Version information:**
 - OS: Pop!_OS 20.10
 - Tailscale version: Tip of `main`

**Comment by josharian** [comment]
cc @DentonGentry who I believe recently looked into DNS-over-TCP

**Comment by DentonGentry** [comment]
We're likely to work on a DNS-over-TLS implementation for https://github.com/tailscale/tailscale/issues/915, supporting DNS on TCP also seems quite reasonable and could be done.

**Comment by danderson** [comment]
Discussed with Adrian, he'll be doing this during the 1.10 cycle.

**Comment by sailorfrag** [comment]
Current status: with #2065, the maximum UDP size is now 4095 bytes which is hopefully larger than the vast majority of DNS responses. Larger UDP responses are possible in the presence of an EDNS record in the client's original request. If a received response is truncated, we now set the header field that signals a truncated packet (which will probably trigger a fallback to TCP that we still don't support yet). The post-truncation packet is still not formed correctly and will confuse some resolvers; #2067 covers that issue. After #2066 is resolved, we will likely stop seeing packets truncated by the local kernel because the upstream DNS server should know not to send packets larger than our maximum, which will mitigate most of the problem.

This issue is being left open to track TCP support, which is more complicated but necessary for truly widespread compatibility.

**Comment by sailorfrag** [comment]
UDP resolution of up to 4095 byte payloads is in v1.10, including setting the truncation bit in the header but not cleaning up any mangled record at the end. EDNS clamping to 4095 bytes (which should prevent uncontrolled truncation leading to a mangled record at the end) will be in the next unstable release (and eventually, v1.12).

Sadly, I must move on to a new task and TCP support will have to wait a little longer; it looks like a moderately large effort (it may require a significantly more ambitious DNS forwarding subsystem) but hopefully the EDNS improvements should mitigate most of the need for TCP.

**Comment by ignoramous** [comment]
Until such a time DNS over TCP can be supported, for folks using tailscale's MagicDNS, shouldn't port 53 on TCP be simply firewalled to prevent leaks?
